opm/simulators/linalg/cuistl/detail/cublas_wrapper.hpp

```diff
@@ -0,0 +1,168 @@
+/*
+  Copyright 2022-2023 SINTEF AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/**
+ * Contains wrappers to make the CuBLAS library behave as a modern C++ library with function overlading.
+ *
+ * In simple terms, this allows one to call say cublasScal on both double and single precisision,
+ * instead of calling cublasDscal and cublasSscal respectively.
+ */
+
+#ifndef OPM_CUBLASWRAPPER_HEADER_INCLUDED
+#define OPM_CUBLASWRAPPER_HEADER_INCLUDED
+#include <cublas_v2.h>
+#include <opm/common/ErrorMacros.hpp>
+
+namespace Opm::cuistl::detail
+{
+
+inline cublasStatus_t
+cublasScal(cublasHandle_t handle,
+           int n,
+           const double* alpha, /* host or device pointer */
+           double* x,
+           int incx)
+{
+    return cublasDscal(handle,
+                       n,
+                       alpha, /* host or device pointer */
+                       x,
+                       incx);
+}
+
+inline cublasStatus_t
+cublasScal(cublasHandle_t handle,
+           int n,
+           const float* alpha, /* host or device pointer */
+           float* x,
+           int incx)
+{
+    return cublasSscal(handle,
+                       n,
+                       alpha, /* host or device pointer */
+                       x,
+                       incx);
+}
+
+inline cublasStatus_t
+cublasScal([[maybe_unused]] cublasHandle_t handle,
+           [[maybe_unused]] int n,
+           [[maybe_unused]] const int* alpha, /* host or device pointer */
+           [[maybe_unused]] int* x,
+           [[maybe_unused]] int incx)
+{
+    OPM_THROW(std::runtime_error, "cublasScal multiplication for integer vectors is not implemented yet.");
```

Comment:
Is there any value in including these integer overloads? The pointers don't convert in any case so there is no risk of accidentally passing arrays other than float or double.

Reply:
We instantiate the `CuVector` with an `int` template argument, so there is a possibility of getting to this part of the code. So for compilation technical reasons, I think this needs to be there.

I think long term `int` multiplication can have a value for some algorithms, hence I am leaving it open now.

Reply:
Do the corresponding integer algorithms exist in cublas?

Reply:
No, I think it only supports vector spaces and not modules, unfortunately. We would have to write them ourselves, but the point is that once we write this (basic arithmetic operation kernels) ourselves, we would get them for whatever raw type we would want (since they can be templated as done in `vector_operations.cu`), hence the cublas dependency could be removed entirely.

I anyway found the current temporary solution easier/cleaner than:

a) overloading every arithmetic method in `CuVector<int>` to throw an error (I don't think `static_assert` would be possible in this case, since we need to instantiate the vector for `int`). There are more arithmetic operations in `CuVector<T>` than there are cublas-routines we have to overload.

b) adding a bunch of tests to the methods of `CuVector<T>` of the form

```C++
if (std::is_same_v<T, int>) { throw_some_exception(); }
```

A long-term solution I think is a mixture of:

1) Implement the corresponding algorithms for `int` (or rather `template<class T>`) akin to what is done already in `vector_operations.cu` if the need for these algorithms arises. If the hand-written algorithms prove to have equal or better performance for our use cases than CuBlas (maybe there is tuning that we can do, but I would not think so) then we can consider also using these algorithms for `float` and `double`. 

2) Add a new `CuBuffer<T>` class whose only job is to hold data (so `CuVector<T>` without the arithmetic expressions and conversions to `Dune::BlockVector`), and let `CuVector<T>` use this through composition). This would provide a data-holder (as in a GPU version of `std::vector`) on the GPU without any arithmetic operations (which is how `CuVector<int>` is used currently in the code), it would also facilitate having a data holder for raw types for which there are no suitable arithmetic operations (a `union` for instance). 

With such a buffer, the interfaces to the functions in `vector_operations.hpp` could also be simplified to only take a `CuBuffer<T>` (and one could add a corresponding `CuBufferView<T>` that would be CUDA kernel compatible (see for instance what I did in Alsvinn with [View.hpp](https://github.com/alsvinn/alsvinn/blob/master/alsfvm/include/alsfvm/memory/View.hpp), instead of a raw memory pointer and length.

Now, I think 2) should be done down the line anyway, but I also think that for the current state of the code, it would only obfuscate and make things more complicated (more classes to reason about) than it would clarify.

Reply:
I agree with both: That 2) is a good solution, and also not required immediately.